Search with k-NN filters (#1814)
opensearch-project/documentation-website: _search-plugins/knn/filter-search-knn.md

---
layout: default
title: Search with k-NN filters
nav_order: 10
parent: k-NN
has_children: false
has_math: true
---

# Search for k-NN with filters
Introduced 2.4
{: .label .label-purple }

You can create custom filters using Query DSL to search for k-nearest neighbors to a query point across an index of vectors. You define the filter criteria within the `knn_vector` data type description in your query using Query DSL.

## How does a k-NN filter work?

> The OpenSearch k-NN plugin uses the Lucene engine version 9.1 that provides a search process for filtered documents that uses a Hierarchical Navigable Small World (HSNW) algorithm to represent a multi-layered graph to filter searches. After a filter is applied to a set of documents to be searched, the algorithm decides whether to perform pre-filtering for an exact kNN search or modified post-filtering for approximate search. The approximate search with filtering guarantees the top number of closest vectors in result.

Reviewer: I'm not sure we need to put 9.1 as Lucene version. It follows the Lucene version from code OpenSearch and for 2.4 it's Lucene 9.4. Maybe change it to reflect initial k-NN/core version that supports Lucene, something like "starting from version 2.2 OpenSearch k-NN plugin uses the Lucene engine that provides ...."

Lucene also provides the capability to operate its `KnnVectorQuery` over a subset of documents. To learn more about Lucene’s new capability, see the [Apache Lucene Documentation](https://issues.apache.org/jira/browse/LUCENE-10382).

### Filtered search performance

The Lucene engine and HSNW algorithm allow you to to apply k-NN searches more efficiently both in terms of relevancy of search results and performance. Consider that if you do an exact search on a large data set, the results are slow, and post-filtering does not guarantee the required number of results you specify for the `k` value.
With this new capability, you can create an approximate k-NN search and apply filters, with the amount of results you need.

### Requirement: Lucene engine with HNSW method

To run k-NN queries with a filter, the Lucene search engine and HSNW method are required. When you create the mapping, you specify these items in the `knn_vector` field description as follows:

```json
"type": "knn_vector",
                "dimension": 2,
                "method": {
                    "name": "hnsw",
                    "space_type": "l2",

>                     "engine": "lucene"

Reviewer: please make sure it's formatted properly,
```
"my_field": {
                "type": "knn_vector",
                "dimension": 2,
                "method": {
                    "name": "hnsw",
                    "space_type": "l2",
                    "engine": "lucene"
                }
            }
```

---
layout: default
title: Search with k-NN filters
nav_order: 10
parent: k-NN
has_children: false
has_math: true
---

# Search for k-NN with filters
Introduced 2.4
{: .label .label-purple }

> You can create custom filters using Query DSL search options to refine your k-NN searches. You define the filter criteria within the `knn_vector` field's `filter` subsection in your query. You can use all available Query DSL query types such as: term, range, regex and wildcard. To include or exclude results, you specify Boolean query clauses. You also specify a query point with the `knn_vector` type and search for nearest neighbors that match your filter criteria.

Reviewer: This looks more like exhaustive list of query types, I think we need to highlight that those four types are only examples, although probably they are ones used the most. I don't think we need to list all types as there are way too many, and also it can be something custom that we're can't predict, like 3-rd party plugin. 
```
You can use all available Query DSL query types such as: term, range, regex and wildcard
```

To run k-NN queries with a filter, the Lucene search engine and HSNW method are required.

To learn more about how to use Query DSL Boolean query clauses, see [Boolean queries]({{site.url}}{{site.baseurl}}/opensearch/query-dsl/bool).
{: .note }

## How does a k-NN filter work?

> Starting from OpenSearch version 2.2 the k-NN plugin uses the Lucene engine that provides a search process for filtered documents that uses a Hierarchical Navigable Small World (HSNW) algorithm to represent a multi-layered graph to filter searches. After a filter is applied to a set of documents to be searched, the algorithm decides whether to perform pre-filtering for an exact kNN search or modified post-filtering for approximate search. The approximate search with filtering guarantees the top number of closest vectors in result.

Reviewer: I think first sentence is a bit misleading in terms of version and products. Lucene introduced knn search in 9.0 and filtering for this search in 9.1. On our end we released lucene based knn in 2.2 (at this point core OpenSearch and knn plugin were using Lucene 9.2). Filtering feature in knn plugin will be released in version 2.4 and it uses Lucene 9.4. 
I think we need to change sentence to something like, please feel free to change exact wording as needed:
```
OpenSearch k-NN plugin uses Lucene engine for search process starting from version 2.2. Lucene engine provides a search that is based on Hierarchical Navigable Small World (HSNW) algorithm to represent a multi-layered graph. From OpenSearch version 2.4 k-NN plugin is able to incorporate filters in search, that is based on Lucene 9.4.
```
and then continue about the filtering algorithm.

Lucene also provides the capability to operate its `KnnVectorQuery` over a subset of documents. To learn more about Lucene’s new capability, see the [Apache Lucene Documentation](https://issues.apache.org/jira/browse/LUCENE-10382).

### Filtered search performance

> The Lucene engine and HSNW algorithm allow you to to apply k-NN searches more efficiently both in terms of relevancy of search results and performance. Consider that if you do an exact search on a large data set, the results are slow, and post-filtering does not guarantee the required number of results you specify for the `k` value.

Reviewer: I think we don't need a separate reference to HSNW algorithm as this is used mainly for search itself and not filtering. We can say something around "Filtering that is tightly integrated with Lucene HNSW algorithm implementation allows you to to apply k-NN searches more efficiently...."

With this new capability, you can create an approximate k-NN search and apply filters, with the amount of results you need.

The following workflow diagram shows how the HSNW algorithm decides which type of filtering to apply to a search based on the volume of documents, and number of `k` points in the index that you search with a filter.

> ![How the algorithm evaluates a doc set]({{site.url}}{{site.baseurl}}/images/hsnw-algorithm.png)

Reviewer: Maybe it makes sense to add a legend for the workflow diagram

```
- N documents in the index
- P documents in the search set after the filter is applied. P <= N
- q is the search vector
- k maximum number of vectors in response
```

***Figure 1: Filter algorithm workflow***

## Filter approaches by use case

Depending on the data set that you are searching, you might choose a different approach to minimize recall or latency. You can create filters that are either: very selective (80%), somewhat selective (38%), or not very selective (2.5%). The selectiveness percentage indicates the amount the filter returns for any given document set in an index.

> In this context, `score_script` is essentially a brute force search, whereas boolean filter is an approximate k-NN search with post filtering.

Reviewer: I think we may need to add more context around `scoring script` and `boolean filter` approaches. They both are available for users today.For scoring script we can provide a link https://opensearch.org/docs/latest/search-plugins/knn/knn-score-script/. I can't find existing link for boolean query post filtering, maybe we can say that in such case knn query joined with a filter through `bool` `must` clause, or provide an example of such query:
```
{
    "size": 3,
    "query": {
        "bool": {
            "filter": {
                "bool": {
                    "must": [
                        {
                            "range": {
                                "rating": {
                                    "gte": 8,
                                    "lte": 10
                                }
                            }
                        },
                        {
                            "term": {
                                "parking": "true"
                            }
                        }
                    ]
                }
            },
            "must": [
                {
                    "knn": {
                        "location": {
                            "vector": [
                                5.0,
                                4.0
                            ],
                            "k": 20
                        }
                    }
                }
            ]
        }
    }
}
```

---
layout: default
title: Search with k-NN filters
nav_order: 10
parent: k-NN
has_children: false
has_math: true
---

# Search for k-NN with filters
Introduced 2.4
{: .label .label-purple }

> You can create custom filters using Query DSL search options to refine your k-NN searches. You define the filter criteria within the `knn_vector` field's `filter` subsection in your query. The most commonly used Query DSL query types are: term, range, regex and wildcard. To include or exclude results, you specify Boolean query clauses. You also specify a query point with the `knn_vector` type and search for nearest neighbors that match your filter criteria.

Reviewer: `The most commonly used Query DSL query types are: term, range, regex and wildcard.`
I think we better change this to point out that any of the registered query types can be used. Maybe something like:
`You can use any of DSL query types that are registered in the OpenSearch as a filter. This includes but not limited to existing OpenSearch types like term, range, regex or wildcard, as well as custom query types`

Reviewer: main idea is that we do support all standard types including most popular ones, but also any other custom one

Author: ok i can change that.

With this new capability, you can create an approximate k-NN search and apply filters, with the amount of results you need.

The following workflow diagram shows how the HSNW algorithm decides which type of filtering to apply to a search based on the volume of documents, and number of `k` points in the index that you search with a filter. The variables shown in the diagram are described in the table below.

Variable | Description |
-- | -- | -- |
N | Number of documents in the index.
P | Number of documents in the search set after the filter is applied using the formula: P <= N.
q | The search vector.
k | The maximum number of vectors to return in the response.

![How the algorithm evaluates a doc set]({{site.url}}{{site.baseurl}}/images/hsnw-algorithm.png)

***Figure 1: Filter algorithm workflow***

## Filter approaches by use case

> Depending on the data set that you are searching, you might choose a different approach to minimize recall or latency. You can create filters that are either: very selective (80%), somewhat selective (38%), or not very selective (2.5%). The selectiveness percentage indicates the amount the filter returns for any given document set in an index.

Reviewer: now I'm not sure selective is good word to describe the filter. Initially I was thinking that more selective filter means more criteria in filter and as a result less documents returned, but seem that can be interpreted in an opposite way and that makes sense as well. 
Maybe we can change it to something more obvious, like strict or restrictive. More strict filter return less document and vise versa.

Author: ok. let me do the rewrites now to change terms throughout as follows: 

very selective -> Restrictive
somewhat selective -> Somewhat restrictive
not very selective -> Not very restrictive

Reviewer: I think it should be the other way around:

very selective -> Not very restrictive
somewhat selective -> Somewhat restrictive
not very selective -> very restrictive

so very restrictive filter leaves the smallest number of docs, 2.5 % in our example, and not restrictive filter has very few criteria and thus retrieves the highest number of docs 80% in our example.

> Depending on the data set that you are searching, you might choose a different approach to minimize recall or latency. You can create filters that are either: very restrictive (80%), somewhat restrictive (38%), or not very restrictive (2.5%). The restrictive percentage indicates the amount of documents the filter returns for any given document set in an index.

Author: hi @martin-gaievski , my apologies. . .looks like i applied the terms to the opposite percentages. Thanks for catching that! making updates now.

#### Filter selectiveness with latency per doc set volume

Number of vectors | Selectiveness of filter, % | k | Recall | Latency
-- | -- | -- | -- | --
10M | 2.5 | 100 | score_script | score_script
10M | 38 | 100 | lucene_filtering | Boolean filter
10M | 80 | 100 | score_script | lucene_filtering
1M | 2.5 | 100 | lucene_filtering | score_script
1M | 38 | 100 | lucene_filtering | lucene_filtering / score_script
1M | 80 | 100 | Boolean filter | lucene_filtering

> In this context, `score_script` is essentially a brute force search, whereas a Boolean filter is an approximate k-NN search with post-filtering.

Reviewer: maybe we need to change `score_script` to Scoring script, this way both Scoring script and Boolean filter will look identical. Essentially in this context we want to refer both in the same manner.

                                        "gte": 5
                                    }
                                }
                            }
                        ]
                    }
                }
```

## Overview: How to use filters in a k-NN search

The workflow to search with a filter includes three steps:
1. Create an index and specify the requirements for Lucene engine and HNSW in the mapping.
1. Add your data to the index.
1. Search the index and specify these three items in your query:
* One or more filters defined by Query DSL
* A vector reference point defined by the `vector` field.
* The number of matches you want returned with the `k` field.

We use a range query to specify hotel feedback ratings and a term query to require that parking is available. The criteria is processed with Boolean clauses to indicate whether or not the document contains the criteria.

Consider a dataset that contains 12 documents, a search reference point, and documents that meet two filter criteria.

![Graph of documents with filter criteria]({{site.url}}{{site.baseurl}}/images/knn-two-filters.png)

## Step 1: Create a new index with a Lucene mapping

> Before you can do a k-NN search with a filter, you need to create an index, specify the Lucene engine in a mapping, and add data to the index.

Reviewer 2: ```suggestion
Before you can run a k-NN search with a filter, you need to create an index, specify the Lucene engine in a mapping, and add data to the index.
```

> You need to add a `location` field to represent the location, and specify it as the `knn_vector` type. The most basic vector can be two dimensions. For example:

Reviewer 2: ```suggestion
You need to add a `location` field to represent the location and specify it as the `knn_vector` type. The most basic vector can be two dimensions. For example:
```

Reviewer 2: Do we mean "can have two dimensions" or "can be two-dimensional"?

```
  "type": "knn_vector",
  "dimension": 2,
```

### Requirement: Lucene engine with HNSW method

Make sure to specify "hnsw" method and "lucene" engine in the `knn_vector` field description, as follows:

```json
"my_field": {
                "type": "knn_vector",
                "dimension": 2,
                "method": {
                    "name": "hnsw",
                    "space_type": "l2",
                    "engine": "lucene"
                }
            }
```

#### Sample request

> The following request creates a new index called "hotels-index."

Reviewer 2: ```suggestion
The following request creates a new index called "hotels-index":
```

```json
PUT /hotels-index
{
    "settings": {
        "index": {
            "knn": true,
            "knn.algo_param.ef_search": 100,
            "number_of_shards": 1,
            "number_of_replicas": 0
        }
    },
    "mappings": {
        "properties": {
            "location": {
                "type": "knn_vector",
                "dimension": 2,
                "method": {
                    "name": "hnsw",
                    "space_type": "l2",
                    "engine": "lucene",
                    "parameters": {
                        "ef_construction": 100,
                        "m": 16
                    }
                }
            }
        }
    }
}
```
#### Sample response

> Upon success, you should receive "200-OK" status with the following response:

Reviewer 2: Should "a" or "the" precede "200-OK"?

```json
{
  "acknowledged" : true,
  "shards_acknowledged" : true,
  "index" : "hotels-index"
}
```

## Step 2: Add data to your index

Next, add data to your index with a PUT HTTP request. Make sure that the search criteria is defined in the body of the request.

#### Sample request

> The following request adds twelve hotel documents that contain criteria such as feedback ratings and whether or not parking is available.  

Reviewer 2: ```suggestion
The following request adds 12 hotel documents that contain criteria such as feedback ratings and whether or not parking is available:  
```

{ "index": { "_index": "hotels-index", "_id": "3" } }
{ "location": [4.9, 3.4], "parking" : "true", "rating" : 9 }
{ "index": { "_index": "hotels-index", "_id": "4" } }
{ "location": [4.2, 4.6], "parking" : "false", "rating" : 6}
{ "index": { "_index": "hotels-index", "_id": "5" } }
{ "location": [3.3, 4.5], "parking" : "true", "rating" : 8 }
{ "index": { "_index": "hotels-index", "_id": "6" } }
{ "location": [6.4, 3.4], "parking" : "true", "rating" : 9 }
{ "index": { "_index": "hotels-index", "_id": "7" } }
{ "location": [4.2, 6.2], "parking" : "true", "rating" : 5 }
{ "index": { "_index": "hotels-index", "_id": "8" } }
{ "location": [2.4, 4.0], "parking" : "true", "rating" : 8 }
{ "index": { "_index": "hotels-index", "_id": "9" } }
{ "location": [1.4, 3.2], "parking" : "false", "rating" : 5 }
{ "index": { "_index": "hotels-index", "_id": "10" } }
{ "location": [7.0, 9.9], "parking" : "true", "rating" : 9 }
{ "index": { "_index": "hotels-index", "_id": "11" } }
{ "location": [3.0, 2.3], "parking" : "false", "rating" : 6 }
{ "index": { "_index": "hotels-index", "_id": "12" } }
{ "location": [5.0, 1.0], "parking" : "true", "rating" : 3 }
```

#### Sample response

> Upon success, you should receive "200-OK" status with entries for each document ID added to the index. The following response is truncated to only show one document:

Reviewer 2: Should "a" or "the" precede "200-OK"?

```json
{
  "took" : 140,
  "errors" : false,
  "items" : [
    {
      "index" : {
        "_index" : "hotels-index",
        "_id" : "1",
        "_version" : 2,
        "result" : "updated",
        "_shards" : {
          "total" : 1,
          "successful" : 1,
          "failed" : 0
        },
        "_seq_no" : 12,
        "_primary_term" : 3,
        "status" : 200
      }
    }
  ]
}

```

## Step 3: Search your data with a filter

> Now you can create a k-NN search that specifies filters using Query DSL Boolean clauses. You need to include your reference point to search for nearest neighbors. Provide an x-y coordinate for the point with the `vector` field, such as `"vector": [ 5.0, 4.0]`

Reviewer 2: Instead of just "using", do we mean "by using"? Otherwise, this can be read as though the filters are using the clauses. Instead of "with", do we mean "within"? Also, it looks like we need a terminating period.

 To learn more about how to specify ranges with Query DSL, see [Range query]({{site.url}}{{site.baseurl}}/opensearch/query-dsl/term/#range-query).
{: .note }

#### Sample request

> The following request creates a k-NN query that only returns the top hotels rated between 8 and 10 and that provide parking. The filter criteria is indicated with the Query DSL `range` query clause to indicate the range for the feedback ratings and a `term` query clause to indicate "parking."

Reviewer 2: Would "...the top hotels rated between 8 and 10 that also provide parking" work? Also, I would probably try to avoid "is indicated...to indicate".

> The following request creates a k-NN query that only returns the top hotels rated between 8 and 10 and that provide parking. The filter criteria is indicated with the Query DSL `range` query clause to indicate the range for the feedback ratings and a `term` query clause to indicate "parking."

Reviewer 2: ```suggestion
The following request creates a k-NN query that only returns the top hotels rated between 8 and 10 and that provide parking. The filter criteria is indicated with the Query DSL `range` query clause to indicate the range for the feedback ratings and a `term` query clause to indicate "parking":
```

The following response indicates that only three hotels met the filter criteria:

```json
{
  "took" : 47,
  "timed_out" : false,
  "_shards" : {
    "total" : 1,
    "successful" : 1,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : {
      "value" : 3,
      "relation" : "eq"
    },
    "max_score" : 0.72992706,
    "hits" : [
      {
        "_index" : "hotels-index",
        "_id" : "3",
        "_score" : 0.72992706,
        "_source" : {
          "location" : [
            4.9,
            3.4
          ],
          "parking" : "true",
          "rating" : 9
        }
      },
      {
        "_index" : "hotels-index",
        "_id" : "6",
        "_score" : 0.3012048,
        "_source" : {
          "location" : [
            6.4,
            3.4
          ],
          "parking" : "true",
          "rating" : 9
        }
      },
      {
        "_index" : "hotels-index",
        "_id" : "5",
        "_score" : 0.24154587,
        "_source" : {
          "location" : [
            3.3,
            4.5
          ],
          "parking" : "true",
          "rating" : 8
        }
      }
    ]
  }
}

```

## Additional complex filter query

> Depending on how restrictive you want your filter to be, you can add multiple query types to a single request, such as `term`, `wildcard`, `regexp`, and `range`. You can then filter out the search results with the Boolean clauses `must`, `should`, and `must_not`.

Reviewer 2: ```suggestion
Depending on how restrictive you want your filter to be, you can add multiple query types to a single request, such as `term`, `wildcard`, `regexp`, or `range`. You can then filter out the search results with the Boolean clauses `must`, `should`, and `must_not`.
```

#### Sample request

> The following request returns hotels that provide parking. This request illustrates alternate mechanisms to obtain the parking filter criteria. It uses a regular expression for the value `true`, a term query for the key-value pair `"parking":"true"`, a wildcard for the characters that spell "true", and the `must_not` clause to eliminate hotels with "parking" set to `false`.

Reviewer 2: ```suggestion
The following request returns hotels that provide parking. This request illustrates alternative mechanisms for obtaining the parking filter criteria. It uses a regular expression for the value `true`, a term query for the key-value pair `"parking":"true"`, a wildcard for the characters that spell "true", and the `must_not` clause to eliminate hotels with "parking" set to `false`:
```

Reviewer 2: Are we illustrating multiple alternative mechanisms or just one? If the latter, "This request illustrates an alternative mechanism..."

                            }
                        },
                        "should": [
                            {
                                "term": {
                                    "parking": "true"
                                }
                            },
                            {
                                "wildcard": {
                                    "parking": {
                                        "value": "t*e"
                                    }
                                }
                            },
                            {
                                "regexp": {
                                    "parking": "[a-zA-Z]rue"
                                }
                            }
                        ],
                        "must_not": [
                            {
                                "term": {
                                    "parking": "false"
                                }
                            }
                        ],
                        "minimum_should_match": 1
                    }
                }
            }
        }
    }
} 
```
#### Sample response

> The following response indicates a few hits for the search with filters:

Reviewer 2: Instead of "hits", can we use "results"?

---
layout: default
title: Search with k-NN filters
nav_order: 15
parent: k-NN
has_children: false
has_math: true
---

# Search for k-NN with filters
Introduced 2.4
{: .label .label-purple }

You can create custom filters using Query DSL search options to refine your k-NN searches. You define the filter criteria within the `knn_vector` field's `filter` subsection in your query. You can use any of the OpenSearch Query DSL query types as a filter. This includes, but is not limited to the common query types: `term`, `range`, `regexp`, `wildcard`, as well as custom query types. To include or exclude results, use Boolean query clauses. You also specify a query point with the `knn_vector` type and search for nearest neighbors that match your filter criteria.

> To run k-NN queries with a filter, the Lucene search engine and HSNW method are required.

Author: great, thanks!

With this new capability, you can create an approximate k-NN search, apply filters, and get the number of results that you need. To learn more about approximate searches, see [Approximate k-nn search]({{site.url}}{{site.baseurl}}/opensearch/search-plugins/knn/approximate-knn/).

The HSNW algorithm decides which type of filtering to apply to a search based on the volume of documents, and number of `k` points in the index that you search with a filter.

![How the algorithm evaluates a doc set]({{site.url}}{{site.baseurl}}/images/hsnw-algorithm.png)

| Variable | Description |
-- | -- | -- |
N | Number of documents in the index.
P | Number of documents in the search set after the filter is applied using the formula: P <= N.
q | The search vector.
k | The maximum number of vectors to return in the response.

To learn more about k-NN performance tuning, see [Performance tuning]({{site.url}}{{site.baseurl}}/opensearch/search-plugins/knn/performance-tuning/).

## Filter approaches by use case

> Depending on the data set that you are searching, you might choose a different approach to minimize recall or latency. You can create filters that are either:

Reviewer 3: Delete the word "either". If you keep it, the bullets need to be revised to use semicolons and the word "or" with the bullet preceding the last bullet. Also, lowercase the word "For".

* Very restrictive — Returns the least number of documents (For example, 2.5%).
* Somewhat restrictive — Returns some documents (For example, 38%).
* Not very restrictive — Returns the most documents (For example, 80%).

The restrictive percentage indicates the amount of documents the filter returns for any given document set in an index.

> Number of vectors | Filter restrictive percentage | k | Recall | Latency

Reviewer 3: Format table headings in Title Case.

-- | -- | -- | -- | --
10M | 2.5 | 100 | Scoring script | Scoring script
10M | 38 | 100 | lucene_filtering | Boolean filter
10M | 80 | 100 | Scoring script | lucene_filtering
1M | 2.5 | 100 | lucene_filtering | Scoring script
1M | 38 | 100 | lucene_filtering | lucene_filtering / Scoring script
1M | 80 | 100 | Boolean filter | lucene_filtering

> In this context, Scoring script is essentially a brute force search, whereas a Boolean filter is an approximate k-NN search with post-filtering.

Reviewer 3: Do we need to format the term "Scoring script" so the reader understands it's referring to the latency name in the table? For example, set it in italics or quotations?

            ]
        }
    }
}
```
#### Sample response

The Boolean query filter returns the following results in the response:

```json
{
  "took" : 95,
  "timed_out" : false,
  "_shards" : {
    "total" : 1,
    "successful" : 1,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : {
      "value" : 5,
      "relation" : "eq"
    },
    "max_score" : 0.72992706,
    "hits" : [
      {
        "_index" : "hotels-index",
        "_id" : "3",
        "_score" : 0.72992706,
        "_source" : {
          "location" : [
            4.9,
            3.4
          ],
          "parking" : "true",
          "rating" : 9
        }
      },
      {
        "_index" : "hotels-index",
        "_id" : "6",
        "_score" : 0.3012048,
        "_source" : {
          "location" : [
            6.4,
            3.4
          ],
          "parking" : "true",
          "rating" : 9
        }
      },
      {
        "_index" : "hotels-index",
        "_id" : "5",
        "_score" : 0.24154587,
        "_source" : {
          "location" : [
            3.3,
            4.5
          ],
          "parking" : "true",
          "rating" : 8
        }
      }
    ]
  }
}
```

> 

Reviewer 3: Do we need the extra lines between 184 and 186?

### Use case 1: Very restrictive 2.5% filter

A very restrictive filter returns the least amount of documents in your data set. For example, the following filter criteria specifies hotels with feedback ratings less than or equal to 3. This 2.5% filter only returns 1 document:

```json
                "filter": {
                    "bool": {
                        "must": [
                            {
                                "range": {
                                    "rating": {
                                        "lte": 3
                                    }
                                }
                            }
                        ]
                    }
                }
```

### Use case 2: Somewhat restrictive 38% filter

> A somewhat restrictive filter returns 38% of the documents in the doc set that you search. For example, the following filter criteria specifies hotels with parking and feedback ratings less than or equal to 8, and returns 5 documents.

Reviewer 3: Delete comma after 8.

```json
               "filter": {
                    "bool": {
                        "must": [
                            {
                                "range": {
                                    "rating": {
                                        "lte": 8
                                    }
                                }
                            },
                            {
                                "term": {
                                    "parking": "true"
                                }
                            }
                        ]
                    }
                }
```

### Use case 3: Not very restrictive 80% filter

> A filter that is not very restrictive will return 80% of the documents that you search. For example, the following filter criteria specifies hotels with feedback ratings greater than or equal to 5, and returns 10 documents.

Reviewer 3: Delete comma after 5.

```json
                "filter": {
                    "bool": {
                        "must": [
                            {
                                "range": {
                                    "rating": {
                                        "gte": 5
                                    }
                                }
                            }
                        ]
                    }
                }
```

## Overview: How to use filters in a k-NN search

The workflow to search with a filter includes three steps:

> 1. Create an index and specify the requirements for Lucene engine and HSNW in the mapping.

Reviewer 3: change HSNW to HNSW

1. Add your data to the index.
1. Search the index and specify these three items in your query:
* One or more filters defined by Query DSL
* A vector reference point defined by the `vector` field.
* The number of matches you want returned with the `k` field.

> We use a range query to specify hotel feedback ratings, and a term query to require that parking is available. The criteria is processed with Boolean clauses to indicate whether or not the document contains the criteria.

Reviewer 3: Delete comma after "ratings"

{ "index": { "_index": "hotels-index", "_id": "3" } }
{ "location": [4.9, 3.4], "parking" : "true", "rating" : 9 }
{ "index": { "_index": "hotels-index", "_id": "4" } }
{ "location": [4.2, 4.6], "parking" : "false", "rating" : 6}
{ "index": { "_index": "hotels-index", "_id": "5" } }
{ "location": [3.3, 4.5], "parking" : "true", "rating" : 8 }
{ "index": { "_index": "hotels-index", "_id": "6" } }
{ "location": [6.4, 3.4], "parking" : "true", "rating" : 9 }
{ "index": { "_index": "hotels-index", "_id": "7" } }
{ "location": [4.2, 6.2], "parking" : "true", "rating" : 5 }
{ "index": { "_index": "hotels-index", "_id": "8" } }
{ "location": [2.4, 4.0], "parking" : "true", "rating" : 8 }
{ "index": { "_index": "hotels-index", "_id": "9" } }
{ "location": [1.4, 3.2], "parking" : "false", "rating" : 5 }
{ "index": { "_index": "hotels-index", "_id": "10" } }
{ "location": [7.0, 9.9], "parking" : "true", "rating" : 9 }
{ "index": { "_index": "hotels-index", "_id": "11" } }
{ "location": [3.0, 2.3], "parking" : "false", "rating" : 6 }
{ "index": { "_index": "hotels-index", "_id": "12" } }
{ "location": [5.0, 1.0], "parking" : "true", "rating" : 3 }
```

#### Sample response

> Upon success, you should receive "200-OK" status with entries for each of the document IDs that you added to the index. The following response is truncated to only show one document:

Reviewer 3: Suggested rewrite: "...for each document ID added to the index."

```json
{
  "took" : 140,
  "errors" : false,
  "items" : [
    {
      "index" : {
        "_index" : "hotels-index",
        "_id" : "1",
        "_version" : 2,
        "result" : "updated",
        "_shards" : {
          "total" : 1,
          "successful" : 1,
          "failed" : 0
        },
        "_seq_no" : 12,
        "_primary_term" : 3,
        "status" : 200
      }
    }
  ]
}

```

## Step 3: Search your data with a filter

Now you can create a k-NN search that specifies filters using Query DSL Boolean clauses. You need to include your reference point to search for nearest neighbors. Provide an x-y coordinate for the point with the `vector` field, such as `"vector": [ 5.0, 4.0]`

 To learn more about how to specify ranges with Query DSL, see [Range query]({{site.url}}{{site.baseurl}}/opensearch/query-dsl/term/#range-query).
{: .note }

#### Sample request

> The following request creates a k-NN query that only returns the top hotels rated between 8 and 10, and that provide parking. The filter criteria is indicated with the Query DSL `range` query clause to indicate the range for the feedback ratings, and a `term` query clause to indicate "parking."

Reviewer 3: Delete comma after 10 and the comma after "ratings"

```json
POST /hotels-index/_search
{
    "size": 3,
    "query": {
        "knn": {
            "location": {
                "vector": [
                    5.0,
                    4.0
                ],
                "k": 3,
                "filter": {
                    "bool": {
                        "must": [
                            {
                                "range": {
                                    "rating": {
                                        "gte": 8,
                                        "lte": 10
                                    }
                                }
                            },
                            {
                                "term": {
                                    "parking": "true"
                                }
                            }
                        ]
                    }
                }
            }
        }
    }
}
```

> #### Sample Response

Reviewer 3: Lowercase "Response"

The following response indicates that only three hotels met the filter criteria:

```json
{
  "took" : 47,
  "timed_out" : false,
  "_shards" : {
    "total" : 1,
    "successful" : 1,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : {
      "value" : 3,
      "relation" : "eq"
    },
    "max_score" : 0.72992706,
    "hits" : [
      {
        "_index" : "hotels-index",
        "_id" : "3",
        "_score" : 0.72992706,
        "_source" : {
          "location" : [
            4.9,
            3.4
          ],
          "parking" : "true",
          "rating" : 9
        }
      },
      {
        "_index" : "hotels-index",
        "_id" : "6",
        "_score" : 0.3012048,
        "_source" : {
          "location" : [
            6.4,
            3.4
          ],
          "parking" : "true",
          "rating" : 9
        }
      },
      {
        "_index" : "hotels-index",
        "_id" : "5",
        "_score" : 0.24154587,
        "_source" : {
          "location" : [
            3.3,
            4.5
          ],
          "parking" : "true",
          "rating" : 8
        }
      }
    ]
  }
}

```

## Additional complex filter query

> Depending on how restrictive you want your filter to be, you can add multiple query types to a single request, such as: `term`, `wildcard`,  `regexp`, and `range`. You can then filter out the search results with the Boolean clauses `must`, `should`, and `must_not`.

Reviewer 3: Delete colon after "as"

To run k-NN queries with a filter, the Lucene search engine and HSNW method are required.

To learn more about how to use Query DSL Boolean query clauses, see [Boolean queries]({{site.url}}{{site.baseurl}}/opensearch/query-dsl/bool).
{: .note }

## How does a k-NN filter work?

The OpenSearch k-NN plugin version 2.2 provided support for the Lucene engine to process k-NN searches. The Lucene engine provides a search that is based on the Hierarchical Navigable Small World (HSNW) algorithm to represent a multi-layered graph. The OpenSearch k-NN plugin version 2.4 is able to incorporate filters for searches based on Lucene 9.4.

> After a filter is applied to a set of documents to be searched, the algorithm decides whether to perform pre-filtering for an exact kNN search or modified post-filtering for approximate search. The approximate search with filtering guarantees the top number of closest vectors in result.

Reviewer 4: "... guarantees the top number of closest vectors in **the results.**" ?

Lucene also provides the capability to operate its `KnnVectorQuery` over a subset of documents. To learn more about Lucene’s new capability, see the [Apache Lucene Documentation](https://issues.apache.org/jira/browse/LUCENE-10382).

### Filtered search performance

> Filtering that is tightly integrated with the Lucene HNSW algorithm implementation allows you to to apply k-NN searches more efficiently both in terms of relevancy of search results and performance. Consider that if you do an exact search on a large data set, the results are slow, and post-filtering does not guarantee the required number of results you specify for the `k` value.

Reviewer 4: "Filtering that is tightly integrated with the Lucene HNSW algorithm implementation allows you to to apply k-NN searches more efficiently, both in terms of relevancy of search results and performance. Consider, for example, an exact search using post-filtering on a large data set that is returning results slowly and not producing the required number of results specified by `k`."
Not sure if that captures the intended meaning.

> With this new capability, you can create an approximate k-NN search and apply filters, with the amount of results you need.

Reviewer 4: "Using Lucene's new capabilities, you can create an approximate k-NN search, apply filters, and get the number of results you need."
Again, I'm not sure if this is what you're trying to say. But I think this section is stating that Lucene's latest approximate k-NN search can produce faster results and apply efficient filters without sacrificing a target number of results.

> The following workflow diagram shows how the HSNW algorithm decides which type of filtering to apply to a search based on the volume of documents, and number of `k` points in the index that you search with a filter. The variables shown in the diagram are described in the table below.

Reviewer 4: As soon as you say, "The following workflow diagram..." I think you need to provide the diagram immediately after, rather than introduce a table before it.
1. diagram
2. key for variables.
Can the key be incorporated in the diagram or come right after it?

Author: yes. we added the descriptive table, but makes sense to put it below the diagram. thanks

                                        "gte": 5
                                    }
                                }
                            }
                        ]
                    }
                }
```

## Overview: How to use filters in a k-NN search

The workflow to search with a filter includes three steps:
1. Create an index and specify the requirements for Lucene engine and HSNW in the mapping.
1. Add your data to the index.
1. Search the index and specify these three items in your query:
* One or more filters defined by Query DSL
* A vector reference point defined by the `vector` field.
* The number of matches you want returned with the `k` field.

We use a range query to specify hotel feedback ratings, and a term query to require that parking is available. The criteria is processed with Boolean clauses to indicate whether or not the document contains the criteria.

Consider a data set that contains 12 documents, a search reference point, and documents that meet two filter criteria.

![Graph of documents with filter criteria]({{site.url}}{{site.baseurl}}/images/knn-two-filters.png)

***Figure 2: Graph of documents that meet filter criteria***

> ## Step 1: Create a new index with a lucene mapping

Reviewer 4: Is Lucene capped in this heading and the sentence that follows?

The following response indicates that only three hotels met the filter criteria:

```json
{
  "took" : 47,
  "timed_out" : false,
  "_shards" : {
    "total" : 1,
    "successful" : 1,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : {
      "value" : 3,
      "relation" : "eq"
    },
    "max_score" : 0.72992706,
    "hits" : [
      {
        "_index" : "hotels-index",
        "_id" : "3",
        "_score" : 0.72992706,
        "_source" : {
          "location" : [
            4.9,
            3.4
          ],
          "parking" : "true",
          "rating" : 9
        }
      },
      {
        "_index" : "hotels-index",
        "_id" : "6",
        "_score" : 0.3012048,
        "_source" : {
          "location" : [
            6.4,
            3.4
          ],
          "parking" : "true",
          "rating" : 9
        }
      },
      {
        "_index" : "hotels-index",
        "_id" : "5",
        "_score" : 0.24154587,
        "_source" : {
          "location" : [
            3.3,
            4.5
          ],
          "parking" : "true",
          "rating" : 8
        }
      }
    ]
  }
}

```

## Additional complex filter query

> Depending on how restrictive you want your filter to operate, you can add multiple query types to a single request, such as: `term`, `wildcard`,  `regexp`, and `range`. You can then filter out the search results with the Boolean clauses `must`, `should`, and `must_not`.

Reviewer 4: "Depending on how restrictive you want your filter, you can add multiple query types..."

#### Sample request

> The following request filters returns hotels that provide parking. This request illustrates alternate mechanisms to obtain the parking filter criteria. It uses a regular expression for the value `true`, a term query for the key-value pair `"parking":"true"`, a wildcard for the characters that spell "true", and the `must_not` clause to eliminate hotels with "parking" set to `false`.

Reviewer 4: "request filter" singular?
